iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnTensors.cpp

```diff
@@ -297,23 +299,39 @@ struct ConvertHLOToLinalgOnTensorsPass
     MLIRContext *context = &getContext();
 
     auto typeConverter = mhlo::createHloToLinalgSignedIntegerConverter();
-    populateHLOToLinalgOnTensorsConversionPatterns(context, *typeConverter,
-                                                   patterns);
+    if (directHloClientLowering) {
+      // NOTE: not using corresponding setupHLOToFlowPatterns because the entire
+      // HLO dialects are marked illegal by this pass.
+      // TODO: Wonky.
+      populateHLOToFlowPatterns(context, patterns);
+      chlo::PopulateDecomposeChloPatterns(context, &patterns);
+      populateHLOBroadcastingToLinalgPatterns(context, *typeConverter,
+                                              patterns);
+      populateHLOToLinalgOnTensorsConversionPatterns(context, *typeConverter,
+                                                     patterns);
+    } else {
+      // Legacy: assumes that HLO client -> HLO has been run previously.
+      populateHLOToLinalgOnTensorsConversionPatterns(context, *typeConverter,
```

Comment:
Why not remove the else statement and previous `populateHLOToLinalgOnTensorsConversionPatterns`, then included a non conditional variant? Makes it clearer `directHloClientLowering` is handling the HLO Client lowerings to HLO.